### ISSUE TITLE
Behavior of parse_schema for schema that reference other schema

### ISSUE THREAD

**Issue by cmessick** [fastavro/fastavro#624]
If I use `parse_schema` to load two schema where one schema references the other, I must specify `expand=True` in order to write an avro file that can be read by `fastavro`. The documentation says about the expand option "This format should be considered an output only and not passed in to other reader/writer functions as it does not conform to the avro specification and will likely cause an exception," so I'm not sure if I've stumbled onto a bug or if I'm just doing something wrong.

EDIT: Edited test case and output to add version number

Given two schema, named `foo.bar.avsc` and `foo.example.avsc`

```
{
    "name": "bar",    
    "namespace": "foo",
    "type": "record",
    "fields": [
        {"name": "example", "type": "foo.example"}
    ]
}
```

```
{
    "name": "example",
    "namespace": "foo",
    "type": "record",
    "fields": [
        {"name": "number", "type": "int"}
    ]
}
```

I have three test cases, where in each I load the schema in a different way and then attempt to write an avro file to disk and then read that avro file.

```
import fastavro
import json
import sys
import traceback

print(f'fastavro.__version__ = {fastavro.__version__}')

def test_read(fname):
    with open(fname, 'rb') as fo:
        avro_reader = fastavro.reader(fo)
        for record in avro_reader:
            print(record)
        print(avro_reader.writer_schema)

def test_write(fname, parsed_schema, n):
    payload = [{'example': {'number': n}}]
    with open(fname, 'wb') as fo:
        fastavro.writer(fo, parsed_schema, payload)

###
#
# Test case 1, load schema using schema.load_schema()
# Both foo.bar.avsc and foo.example.avsc are in run directory
#
###

print('\nStart of test case 1')

test1_parsed_schema = fastavro.schema.load_schema('foo.bar.avsc')
test_write('test1.avro', test1_parsed_schema, 1)
test_read('test1.avro')

###
#
# Test case 2, load schema using schema.parse_schema()
#
###

print('\nStart of test case 2')

test2_named_schemas = {}
for schema in ('foo.example.avsc', 'foo.bar.avsc'):
    with open(schema, 'r') as f:
        test2_parsed_schema = fastavro.schema.parse_schema(json.load(f), test2_named_schemas)

test_write('test2.avro', test2_parsed_schema, 2)
try:
    test_read('test2.avro')
except Exception as e:
    t, v, tb = sys.exc_info()
    traceback.print_exception(t, v, tb)

###
#
# Test case 3, load schema using schema.load_schema() with expand=True
# Both foo.bar.avsc and foo.example.avsc are in run directory
#
###

print('\nStart of test case 3')

test3_named_schemas = {}
for schema in ('foo.example.avsc', 'foo.bar.avsc'):
    with open(schema, 'r') as f:
        test3_parsed_schema = fastavro.schema.parse_schema(json.load(f), test3_named_schemas, expand=True)

test_write('test3.avro', test3_parsed_schema, 3)
test_read('test3.avro')
```

Running this produces the following output

```
fastavro.__version__ = 1.5.2

Start of test case 1
{'example': {'number': 1}}
{'type': 'record', 'name': 'foo.bar', 'fields': [{'name': 'example', 'type': {'type': 'record', 'name': 'foo.example', 'fields': [{'name': 'number', 'type': 'int'}]}}]}

Start of test case 2
Traceback (most recent call last):
  File "testcase.py", line 48, in <module>
    test_read('test2.avro')
  File "testcase.py", line 10, in test_read
    avro_reader = fastavro.reader(fo)
  File "fastavro/_read.pyx", line 1056, in fastavro._read.reader.__init__
  File "fastavro/_read.pyx", line 1030, in fastavro._read.file_reader.__init__
  File "fastavro/_schema.pyx", line 92, in fastavro._schema.parse_schema
  File "fastavro/_schema.pyx", line 246, in fastavro._schema._parse_schema
  File "fastavro/_schema.pyx", line 291, in fastavro._schema.parse_field
  File "fastavro/_schema.pyx", line 130, in fastavro._schema._parse_schema
fastavro._schema_common.UnknownType: foo.example

Start of test case 3
{'example': {'number': 3}}
{'type': 'record', 'name': 'foo.bar', 'fields': [{'name': 'example', 'type': {'type': 'record', 'name': 'foo.example', 'fields': [{'name': 'number', 'type': 'int'}]}}]}
```

**Comment by scottbelden** [comment]
It looks like the first example works, right? Using `load_schema` to read in multiple dependent schemas is the way I would expect to handle the two schemas you have.

Option 2 definitely won't work because it won't fix the name references (which is the goal of `load_schema`).

Option 3 happens to work for this case, but could bite you later with different schemas. The documentation should probably be modified so that instead of `as it does not conform to the avro specification and will likely cause an exception` it says `as it can produce schemas that not conform to the avro specification and could cause exceptions`. Where `expand` will cause problems is if you have a schema that references a named schema more than once. In a typical avro schema a named schema must be defined only once and then on out its name must be used. The `expand` would cause the name to be expanded and defined more than once which in this library will cause a `SchemaParseException`.

**Comment by cmessick** [comment]
I included test case 1 just as a sanity check for myself to make sure the normal case worked as expected. As far as I know, I need to use `parse_schema` because I can't control the filenames of the schema I'll be using. 

If I understand your response for option 3, it sounds like this should be okay for my use case where we have one primary schema for our messages that uses records from other schema as types for convenience, i.e. the one "main" schema is the only one that references any of the others.

**Comment by scottbelden** [comment]
Given your schemas (many independent and one main schema that references the others), you could also see if [`load_schema_ordered`](https://fastavro.readthedocs.io/en/latest/schema.html#fastavro._schema_py.load_schema_ordered) might be the right choice. You would just list your main schema it sounds like it should work.

**Comment by cmessick** [comment]
`load_schema_ordered` doesn't work because the filenames don't end in avsc. I'm hosting schema files in one location, and then downloading them using an `astropy` function that will download the schema and and cache them locally. This function doesn't give me any control of the filename, even a suffix, unfortunately.

**Comment by scottbelden** [comment]
Gotcha. Yeah, using filenames may have been a poor choice for the API. Perhaps something like tuple of fullname and a file-like object would have been better).

If your schemas become more complex then there is probably a way to use some recent functionality that allows one to specify how schemas will be read (rather than by the filename convention). It's not documented that well, but a test is here that shows a little bit: https://github.com/fastavro/fastavro/blob/627faa64d335ceaadbe418e23404d1b24225b32b/tests/test_schema.py#L1004-L1043

But at the moment it's probably simpler to go with your original option 3. It sounds like it should work for your schemas.
